File: IsraelHiking.Web/android/app/src/main/java/com/mapeak/car/CarRouteData.kt
package com.mapeak.car

import com.getcapacitor.JSObject
import org.json.JSONException
import org.json.JSONObject
import org.maplibre.android.geometry.LatLng

data class CarRouteData(
        val lngLats: List<LatLng> = emptyList(),
        val weight: Double = 0.0,
        val color: String? = null,
        val opacity: Double = 0.0,
        val name: String? = null,
        val markers: List<CarMarkerData> = emptyList()
) {
    /**
     * The distance in meters from the start of the route to each of its points. Measured once,
     * since it only changes when the route does, while the GPS position is matched against the
     * route on every fix.
     */
    val distancesAlongRouteMeters: DoubleArray by lazy {
        val distances = DoubleArray(lngLats.size)
        for (index in 1 until lngLats.size) {
            distances[index] =
                    distances[index - 1] +
                            SpatialHelper.distanceMeters(lngLats[index - 1], lngLats[index])
        }
        distances
    }

    /** The length of the whole route in meters. */
    val lengthMeters: Double
        get() = distancesAlongRouteMeters.lastOrNull() ?: 0.0

    companion object {
        @Throws(JSONException::class)
        fun fromJson(json: JSONObject): CarRouteData {
            val points = json.getJSONArray("points")
            val lngLats =
                    List(points.length()) { i ->
                        val point = points.getJSONArray(i)
                        LatLng(point.getDouble(1), point.getDouble(0))
                    }
            val markersJson = json.optJSONArray("markers")
            val markers =
                    if (markersJson == null) emptyList()
                    else List(markersJson.length()) { i -> CarMarkerData.fromJson(markersJson.getJSONObject(i)) }
            return CarRouteData(
                    lngLats = lngLats,
                    weight = json.getDouble("weight"),
                    color = json.getString("color"),
                    opacity = json.getDouble("opacity"),
                    name = json.optString("name").ifEmpty { null },
                    markers = markers
            )
        }

        @Throws(JSONException::class)
        fun listFromJson(json: JSObject?): List<CarRouteData> {
            val routes = json?.getJSONArray("routes") ?: return emptyList()
            return List(routes.length()) { i -> fromJson(routes.getJSONObject(i)) }
        }
    }
}

File: IsraelHiking.Web/ios/App/App/Car/CarRouteData.swift
import CoreLocation
import Foundation

/// Mirrors `CarRouteData.kt`: one planned route as fed from JS via `Car.storeValue`.
struct CarRouteData {
    let coordinates: [CLLocationCoordinate2D]
    let weight: Double
    let color: String?
    let opacity: Double
    let markers: [CarMarkerData]

    /// The distance in meters from the start of the route to each of its points. Measured once,
    /// since it only changes when the route does, while the GPS position is matched against the
    /// route on every fix.
    let distancesAlongRouteMeters: [Double]

    /// The length of the whole route in meters.
    var lengthMeters: Double { distancesAlongRouteMeters.last ?? 0 }

    private static func distancesAlongRoute(_ coordinates: [CLLocationCoordinate2D]) -> [Double] {
        var distances = [Double](repeating: 0, count: coordinates.count)
        for index in 1..<max(coordinates.count, 1) {
            distances[index] = distances[index - 1]
                + SpatialService.distanceMeters(coordinates[index - 1], coordinates[index])
        }
        return distances
    }

    /// Parses a single `{ points: [[lng, lat], ...], weight, color, opacity, markers }` object.
    static func from(_ json: [String: Any]) -> CarRouteData? {
        guard let points = json["points"] as? [[Any]] else { return nil }
        let coordinates: [CLLocationCoordinate2D] = points.compactMap { pair in
            guard pair.count >= 2,
                  let lng = (pair[0] as? NSNumber)?.doubleValue,
                  let lat = (pair[1] as? NSNumber)?.doubleValue
            else { return nil }
            return CLLocationCoordinate2D(latitude: lat, longitude: lng)
        }
        let markers = (json["markers"] as? [[String: Any]])?.compactMap(CarMarkerData.from) ?? []
        return CarRouteData(
            coordinates: coordinates,
            weight: (json["weight"] as? NSNumber)?.doubleValue ?? 0,
            color: json["color"] as? String,
            opacity: (json["opacity"] as? NSNumber)?.doubleValue ?? 0,
            markers: markers,
            distancesAlongRouteMeters: distancesAlongRoute(coordinates)
        )
    }

    /// Parses the `{ routes: [...] }` payload stored under `CarStoreKeys.route`.
    static func list(from payload: [String: Any]?) -> [CarRouteData] {
        guard let routes = payload?["routes"] as? [[String: Any]] else { return [] }
        return routes.compactMap { from($0) }
    }
}
